// payload-api/lib/exceptions.js
var utils = require('./utils');

class PayloadError extends Error {
    constructor(description, data) {

        if ( data )
            description += '\n\n'+JSON.stringify(data)

        super(description)

        this.data = data
        if ( this.data && this.data.details )
            this.details = this.data.details

        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, this.constructor);
        } else {
            this.stack = (new Error(description)).stack;
        }
    }

    get http_code() {
        this.constructor.http_code
    }
}

class UnknownResponse extends PayloadError {
    type = 'UnknownResponse'
}

class BadRequest extends PayloadError {
    type = 'BadRequest'
}
BadRequest.http_code = 400

class TransactionDeclined extends BadRequest {
    type = 'TransactionDeclined'
    constructor(description, data) {
        super(description, data)
        var object = utils.get_object_cls(this.details)
        if (object)
            this.details = new object(this.details)
        this.transaction = this.details
    }
}
TransactionDeclined.http_code = 400

class InvalidAttributes extends BadRequest {
    type = 'InvalidAttributes'
}
InvalidAttributes.http_code = 400

class Unauthorized extends PayloadError {
    type = 'Unauthorized'
}
Unauthorized.http_code = 401

class Forbidden extends PayloadError {
    type = 'Forbidden'
}
Forbidden.http_code = 403

class NotFound extends PayloadError {
    type = 'NotFound'
}
NotFound.http_code = 404

class TooManyRequests extends PayloadError {
    type = 'TooManyRequests'
}
TooManyRequests.http_code = 429

class InternalServerError extends PayloadError {
    type = 'InternalServerError'
}
InternalServerError.http_code = 500

class ServiceUnavailable extends PayloadError {
    type = 'ServiceUnavailable'
}
ServiceUnavailable.http_code = 503

module.exports = {
    UnknownResponse: UnknownResponse,
    PayloadError: PayloadError,
    BadRequest: BadRequest,
    TransactionDeclined: TransactionDeclined,
    InvalidAttributes: InvalidAttributes,
    Unauthorized: Unauthorized,
    Forbidden: Forbidden,
    NotFound: NotFound,
    TooManyRequests: TooManyRequests,
    InternalServerError: InternalServerError,
    ServiceUnavailable: ServiceUnavailable
}
